grammar_test/literals.txt

```diff
@@ -0,0 +1,240 @@
+======
+symbol
+======
+
+:foo
+
+---
+
+(program (symbol))
+
+=======
+integer
+=======
+
+1234
+
+---
+
+(program (integer))
+
+=======================
+integer with underscore
+=======================
+
+1_234
+
+---
+
+(program (integer))
+
+===========================
+integer with decimal prefix
+===========================
+
+0d1_234
+
+---
+
+(program (integer))
+
+===============================
+integer with hexadecimal prefix
+===============================
+
+0xa_bcd_ef0_123_456_789
+
+---
+
+(program (integer))
+
+=========================
+integer with octal prefix
+=========================
+
+01234567
+0o1234567
+
+---
+
+(program (integer) (integer))
+
+==========================
+integer with binary prefix
+==========================
+
+0b1_0
+
+---
+
+(program (integer))
+
+=====
+float
+=====
+
+1.234_5e678_90
+
+---
+
+(program (float))
+
+=======
+boolean
+=======
+
+true
+
+---
+
+(program (boolean))
+
+===
+nil
+===
+
+nil
+
+---
+
+(program (nil))
+
+====================
+single-quoted string
+====================
+
+''
+' '
+'  '
+
+---
+
+(program (string) (string) (string))
+
+==============================================
+single-quoted string with escaped single quote
+==============================================
+
+'\''
+
+---
+
+(program (string))
+
+===========================================
+single-quoted string with escaped backslash
+===========================================
+
+'\\'
+
+---
+
+(program (string))
+
+=================================================
+single-quoted string with backslash and character
+=================================================
+
+'\d'
+
+---
+
+(program (string))
+
+====================
+double-quoted string
+====================
+
+""
+" "
+"  "
+
+---
+
+(program (string) (string) (string))
+
+==============================================
+double-quoted string with escaped double quote
+==============================================
+
+"\""
+
+---
+
+(program (string))
+
+===========================================
+double-quoted string with escaped backslash
+===========================================
+
+"\\"
+
+---
+
+(program (string))
+
+=================================================
+double-quoted string with backslash and character
+=================================================
+
+"\d"
+
+---
+
+(program (string))
+
+=============
+interpolation
+=============
+
+"#{foo}"
+
+---
+
+(program (string))
+
+=====================
+escaped interpolation
+=====================
+
+"\#{foo}"
+
+---
+
+(program (string))
+
+========================================
+percent q string with various delimiters
+========================================
+
+%q!a!
+%q@a@
+%q$a$
+%q%a%
+%q^a^
+%q&a&
+%q*a*
+%q-a-
+%q+a+
+%q/a/
+%q|a|
+%q?a?
+%q\a\
+%q.a.
+%q,a,
+%q:a:
+%q;a;
+%q'a'
+%q"a"
+%q)a)
+%q]a]
+%q}a}
+%q>a>
+%q`a`
+%q~a~
+%q_a_
+%q#a#
```

Comment:
This last one only works because we’re able to parse the entire sequence using a single `RegExp`. I don’t think we’re going to be able to make it work for `%` or `%Q` strings, unfortunately, for much the same reason that interpolations are trouble in double quoted strings.